### PR TITLE
Added the macro with-generators.

### PR DESCRIPTION
**Pull request by malisper**
It makes it easier to bind generators to variables and run tests with
them. For example:

```
(with-generators ((x (integer)) 
                  (y (integer)))
  (check-that (= (+ x y) (+ x y)))
  ;; Case designed to fail.
  (check-that (= (+ x y) (+ x y 1))))
```

I plan on writing a more powerful version that would allow something such as the following:

```
(with-generators ((x (integer))
                  (y (integer)))
  (let ((result (+ x y)))
    (check-that (= result result))
    (check-that (= result (+ result 1)))))
```

I can think of no easy way to implement it except by using a code walker.

### CODE REVIEW

**Comment by DalekBaldwin** [comment]
I wanted to use a syntax like this from the beginning, but I thought it might make it harder to see at a glance how other features are interacting with it (in-line examples, regression cases).

And there's a tradeoff with respect to DRY-ness; you can implicitly wrap each definition with `(generator ...)` like you did here but then you wouldn't be able to reuse a generator predefined somewhere else. I've been leaning toward providing a reader macro like `#G(foo)` --> `(generator (foo))`, which would justify getting rid of the implicit `generator` macro in chained generators and in any future features. I think I'll add that next and see if it tightens up the DSL enough to incorporate your macro without too much visual clutter.

**Comment by DalekBaldwin** [comment]
I made a new branch showing what that would look like: https://github.com/DalekBaldwin/check-it/commit/479038665aec23bcae81691320d56caa691a28f8

I like it so far and I'll try it with your macro later.

**Comment by malisper** [comment]
I decided to try out check-it by writing some tests for one of my libraries. There were two main problems I ran in to. The `#G` reader macro solves one of them, which was using the `generator` macro all of the time. The other problem was that I was forced to duplicate the calls to the actual function being tested. This was the main reason I started working on `with-generators`, with the goal of achieving the second version. 

As for the problem you mention with not being able to reuse a predefined generator, I see two ways it can be solved. One way is to pass the generator to `with-generators`, since using `generator` with a non-generator expression winds up being equivalent to just the expression. The other option is to have `check-that` take a keyword argument for passing in a predefined generator.

**Comment by DalekBaldwin** [comment]
"One way is to pass the generator to with-generators, since using generator with a non-generator expression winds up being equivalent to just the expression."

Now that you point it out, I realize the current design allows for that workaround, but the main reason it works that way is so that non-trivial constant values can easily be inserted as alternatives in an `or` generator. Still, I don't want people passing in an expression that they expect to be interpreted as normal Lisp that inadvertently gets processed in the generator DSL, so I prefer the explicitness of `#G`.

Also, the code you submitted in this PR runs a separate set of randomized tests for each invocation of `check-that`, but the way your example for the proposed second version is written, I'd expect that identical values of `result` would be used in both runs of `check-that`. But it sounds like you want it to have the same effect as this:

``` lisp
(let* ((x #G(integer))
       (y #G(integer))
       (result #G(map #'+ x y)))
  (check-it result (lambda (z) (= z z)))
  (check-it result (lambda (z) (= z (+ z 1)))))
```

Code walking wouldn't be necessary if you just used the same generated values for every test in the scope. I don't think there's any reason you'd specifically need to ensure that each test within a scope could potentially receive different values, and it can make a lot of sense to run multiple tests on the exact same generated values, since there can be important information relayed by the fact that one particular value passes some tests and fails others. But once this library starts allowing destructive operations on generated values, value reuse will become trickier (I suppose I could generate a single random state at the top of the block and restore it after each destructive test). So having a construct that lets you scope certain facets of how a generator is used might be pretty useful then.

I tinkered with this for a bit before getting to what I think is the essence of the problem and  the solution. You want to separate the naming of something from how it's used, despite the fact that the name will refer to different values at different times, but in a predictable way. And you don't want to throw in a bunch of extraneous `lambda`s and `destructuring-bind`s. So it's not terribly different from e.g. handling loop iterator variables.

Here's how I'd go about doing it without a code walker. Say your whole expression looks like this:

``` lisp
(with-generators ((x #G(integer))
                  (y #G(integer))
                  (z #G(integer)))
  (with-derived-values ((herp (+ x y))
                        (derp (+ y z)))
    (check-that (= herp derp))))
```

`with-derived-values` would create functions that simply take values corresponding to the active generators from the surrounding context: `(lambda (x y z) (declare (ignorable x y z)) (+ x y))` and `(lambda (x y z) (declare (ignorable x y z)) (+ y z))`. Then each `check-that` form would generate values from the generators and bind them to x, y, and z, then call the two functions and bind the results to herp and derp, then execute its body.

And the binding syntax could be similar to `chain` when you want to reuse a generator already bound to a variable and just reinterpret the binding as a binding to the values it generates:

``` lisp
(let ((x #G(integer))
      (y #G(integer))
      (z #G(integer)))
  (with-generators (x y z)
    ...))
```

You could probably extend this pattern to come up with new ways to write chained generators, locally-scoped user-defined generators, etc., whenever that's a more natural way to express the composite type spec you want.

The Hypothesis docs would make a big deal about this being some kind of monadic bind blah blah. But it's just hiding a bunch of lambda munging. I'll play with it for a bit and see if I come up with something elegant.

**Comment by malisper** [comment]
I actually do want to separate the individual test cases since shrinking becomes almost impossible when multiple tests are lumped together. What you are suggesting for `with-derived-values` may work wrt shrinking. The reason I want to use a code walker is to separate out the individual test cases so that shrinking remains possible. For example:

```
(with-generators ((x (integer))
                  (y (integer)))
  (let ((result (+ x y)))
    (check-that (= result result))
    (check-that (= result (+ result 1)))))
```

would be similar to

```
(with-generators ((x (integer))
                  (y (integer)))
  (let ((result (+ x y)))
    (check-that (= result result)))
  (let ((result (+ x y)))
    (check-that (= result (+ result 1)))))
```

Since there is now no code shared between the two, shrinking remains trivial.  How about we both implement what we are thinking and compare the results?

As for possibly confusing the generator DSL and regular lisp code, a possible fix would be to have the DSL signal an error when it sees an expression that it doesn't recognize. Then having a specific generator syntax, say `(lisp ...)`, that allows the user to write pure lisp code.

**Comment by malisper** [comment]
So I was able to get something really cool working by using a code walker. It is a macro `with-tests` which is allows for multiple check-that forms to use a single variable that depends on generated values. The example I gave for `with-generators` would look something like this using `with-tests`:

```
(with-tests
  (let* ((x (generator (integer)))
         (y (generator (integer)))
         (result (+ x y)))
    (check-that (= result result))
    (check-that (= result (+ result 1)))))
```

One cool thing about this is that your `#G` macro would work with it, since it doesn't handle the binding of generators itself. It also allows for the use of generators defined outside of the `with-tests` form, you just have to call `generate` explicitly to obtain the generated values.

**Comment by DalekBaldwin** [comment]
I put together a little library that lets you bind variables and functions in such a way that you can pretend your macros are expanded recursively, i.e. within the dynamic scope of their parents' expansions: https://github.com/DalekBaldwin/macrodynamics/blob/master/src/macrodynamics.lisp

It's built on the same basic foundation as the compile-time condition/continuation idea from a while back, but it's much simpler since we don't need to communicate any information from inner forms to outer forms.

Here's how I implemented your proposed DSL that way: https://github.com/DalekBaldwin/macrodynamics/blob/master/test/macros.lisp

`with-generators` expands into an ordinary `let`, but also "dynamically" binds metadata for use in macroexpansions further down the code tree.

`let-map` inspects expansion-time metadata about generator bindings to figure out what variables need to be captured in the function definitions it expands into, and it binds some new metadata about those functions too. (I'd prefer not to overload `let` when there's already so much semantic voodoo going on and I don't think I could overload it anyway without using a code walker or requiring client packages to shadow it. You already know my thoughts on code walkers, and even if it's possible to do everything we want with them right now, I imagine it would be harder to integrate other features with them later.)

`check-that` generates a tuple of bound generators like before, plus it wraps the test expression with bindings computed from the mapping functions.

Putting it all together, it works fine with multiple l layers of mappings:

``` lisp
(with-generators ((x #g(integer))
                  (y #g(integer)))
  (let-map ((a (+ x x)))
    (let-map ((b (+ a a)))
      (let ((check-it:*num-trials* 10))
        (check-that (progn
                      (format t "~&~A ~A ~A ~A~%" x y a b)
                      (= a a)))))))

;; output
10 -3 20 40
-6 -7 -12 -24
1 -1 2 4
-6 3 -12 -24
3 2 6 12
6 -3 12 24
7 10 14 28
5 6 10 20
2 3 4 8
5 3 10 20
```

If I modify the generator DSL to hook into expansion-time metadata, I think I could go a step further and create lexically-bound custom recursive generator types too. And then I'd guess chained generators would already work like this:

``` lisp
(with-generators ((x #g(integer)))
  (with-generators ((y #g(chain (x) #g(list (list (integer) :length x)
                                            :length x))))
    (check-that (equal y y))))
```

But then the standard generator DSL is a bit redundant in requiring bindings in the `chain` form, and it would generate two separate values for `x`, one on its own, and one in the process of generating `y`. I can't say whether that would be good or bad until I think up some realistic use cases. I could write another dynenv-macro like so:

``` lisp
(with-generators ((x #g(integer)))
  (chaining ((y #g(list (list (integer) :length x)
                        :length x)))
            (check-that (equal y y))))
```

Then the expansion-time metadata would basically keep track of a dependency graph to avoid shrinking any generators used to parameterize a chained generator and to make sure every lexically-indicated generator is generated only once per test. Who knows, some day there might be a good reason to chain through multiple levels of generators. But I like how this gets rid of all the extraneous rebinding code and makes the basic logic/dataflow pop out at you.

**Comment by DalekBaldwin** [comment]
I improved the helper DSL and submitted it to Quicklisp: https://github.com/quicklisp/quicklisp-projects/issues/982

**Comment by malisper** [comment]
I was using check-it to run some tests on an implementation of red-black trees and realized that there is a problem with the way you are using closures to keep track of the variables - it is no longer possible to use mutation on those variables! I would generate a list of values and then insert them all into the tree while at the same time testing properties of the tree. In the case of failure, the `check-it` expression needs to know how the tree was built. I don't think it is possible to keep track of all of the mutations with closures.

**Comment by DalekBaldwin** [comment]
Is this something to do with normal value generation, or the lexical DSL we're designing here? Can you show me an example of the problem?

**Comment by malisper** [comment]
This has to do with the lexical DSL. The implementation of red-black trees I was testing is [hh-redblack](https://github.com/hargettp/hh-redblack) (I thought there was a bug, but couldn't figure out how to trigger it, so I used check-it to trigger it for me). Here is what similar code would look like without using the DSL:

```
(check-it (generator (list (integer)))
          (lambda (xs)
            (let ((tree (make-red-black-tree)))
              (and (loop for x in xs 
                         do (rb-put tree x "Some data") 
                         always (some-test tree))
                   (loop for x in (shuffle xs) 
                         do (rb-remove x) 
                         always (some-other-test tree))))))
```

The problem with doing the above is that if either of the tests fails, the data will keep being shrunk. It may not be the same test as the one that actually triggered the failure (actually I'm not sure if this is a bad thing. If check-it finds a failed test, does it really matter which test it is?). With the lexical DSL the idea is to share the data across multiple tests, but in such a way that the data can be shrunk with respect to a single test. With my implementation of the DSL (and proper short circuiting, which I just added), I can do something like the following:

```
(with-tests
   (let* ((xs (generator (list (integer))))
         (tree (make-red-black-tree)))
     (loop for x in xs do 
       (rb-put tree x "Some data") 
       (check-that (some-test tree)))
     (loop for x in (shuffle xs) do 
       (rb-remove tree x) 
       (check-that (some-other-test tree)))))
```

The issue is that the tree is being modified as it is being tested. I'm pretty sure there isn't an easy way to rewrite the above code using `let-map` since `check-that` needs to know about the mutation that goes on and redo that mutation every time it reruns the test. Just my thoughts.

Edit: What I'm trying to get at is that the whole point of the DSL is to share the same data across multiple tests. `let-map` is one way to do that. It just doesn't seem like there is an easy way to share side effects across multiple tests through `let-map` or some similar mechanism.

**Comment by DalekBaldwin** [comment]
This looks a lot different from what you were talking about earlier. What do you intend that `with-tests` form to expand into?

**Comment by malisper** [comment]
It roughly expands into the following (`#:` represents a gensym):

```
(let ((#:generator (generator (tuple (list (integer))))))
  (check-it% '(some-test tree) #:generator
             (lambda (#:args)
               (block #:test-block
                 (destructuring-bind (#:generator-value) #:args
                     ;; Now my actual code.
                     (let* ((xs #:generator-value))       
                            (tree (make-red-black-tree)))
                       (loop for x in xs do 
                         (rb-put tree x "Some data") 
                         ;; Only the first check-that is "expanded".
                         (unless (some-test tree) (return-from #:test-block nil)))

                       ;; This code is completely useless.
                       (loop for x in (shuffle xs) do 
                         (rb-remove tree x)) ; The second one is removed

                       ;; If we made it here, we passed all of the tests.
                       t)))))
   ;; Another %check-it that is the same as the above except
   ;; the second check-that is expanded instead of the first.
   (...)) 
```

As I was typing this out, I realized I am needlessly duplicating the code. I could just have a number which designates the current test being ran.

**Comment by DalekBaldwin** [comment]
Okay. At first I understood what you were doing to be taking `let` forms, etc., and shoving them down inside each `check-that` form, so it was a bit confusing to see `check-that` forms inside the loops. But now it looks like what you're really doing is declaring that the code-walker-modified binding form should demarcate the true scope of any single set of randomly-generated values, and then running its body `*size*` times for each `check-that` form that appears lexically inside of it, with a different `check-that` form being actually active on each batch of `*size*` generated values.

So when the form testing `some-other-test` is active, it's being applied to different randomly generated values than the ones that were used in the random batch that tested `some-test`.

Here are my criticisms, hopefully in a coherent order:

As a matter of testing style/philosophy, I'd argue that when you're dealing with this sort of stateful path dependence, it really ought to be treated as a single self-contained test, because otherwise seeing the results of two tests may lull you into a false sense of security that you are seeing two distinct and unentangled pieces of information.

If `rb-put` is buggy but you strip the checking form out of that section of the code and don't short-circuit it there, there's a good chance that if you get a failure in the `rb-remove` loop, it could actually be due to problems in `rb-put` rather than `rb-remove`. And since different values are being generated when each check is active, if you see `some-other-test` fail, you won't know if `some-test` also would have failed, which is what you would probably need to know to easily figure out whether the failure of `some-other-test` actually indicated a bug in `rb-put` rather than `rb-remove`. If your bug only shows up with very rare test inputs, and the user happens to only hit upon one of those inputs when the `some-other-test` form is active, they may mistakenly assume that the `rb-put` part of the code behaved just fine. Even if they're the same person who wrote the test.

Since the underlying `check-it%` function already treats unhandled errors as test failures, if you need to short-circuit the test, you can just short-circuit the test:

``` lisp
(with-tests
  (let* ((xs (generator (list (integer))))
         (tree (make-red-black-tree)))
    (check-that
      (progn
        (loop for x in xs do 
             (rb-put tree x "Some data") 
             (assert (some-test tree)))
        (loop for x in (shuffle xs) do 
             (rb-remove tree x) 
             (assert (some-other-test tree)))
        t))))
```

But say you really, really needed to distinguish failures in the `rb-put` section from failures in the `rb-remove` section in your test reporting protocol. You still couldn't really do that unless you integrated your current version of this DSL more tightly with whatever other test framework you're using; e.g. if you're using Stefil, you can't put each separate call to `check-it%` within an `is` macro, which is what you would need to do to report them as separate self-contained tests. Or at least you'd have to make this DSL more complicated to let the user express a way to do that on their own, which is inconvenient.

And even if you really, really wanted to do this and only let failures shrink with respect to one assertion or the other, it's still extrinsic to the domain-level concerns of a testing tool like check-it, and it would be better off handled by a more generic independent utility for ad-hoc code compression. If you want one `assert` to be active in the first run and a different `assert` to be active in the second run, you could use a code-walking strategy wholly ignorant of check-it's internals to do something like this:

``` lisp
(with-tests
  (let* ((xs (generator (list (integer))))
         (tree (make-red-black-tree)))
    (duplicate-for-each-subform-named derp
      (is (check-that
           (progn
             (loop for x in xs do 
                  (rb-put tree x "Some data") 
                  (derp (assert (some-test tree))))
             (loop for x in (shuffle xs) do 
                  (rb-remove tree x) 
                  (derp (assert (some-other-test tree))))
             t))))))
```

This would be flexible enough to easily create separate `is` tests and also do a better job of alerting anybody reading the code to the weird semantic voodoo going on. (I've occasionally done something like this where, based on the result of some predicate, I needed the basic code template to vary in spots that did not correspond to standalone expressions; otherwise I'd have to insert a `cond` expression up a level or two from where the behavior actually differed and duplicate the rest of the code inside of it.)

So basically my verdict is this: declaratively specifying how variables will eventually be bound to generated values is a valid domain-level concern of property-based testing and probably worth having an integrated DSL for, but playing code golf with idiosyncratic variations on one-off code templates is not.

**Comment by malisper** [comment]
You definitely understand what I am trying to do with the DSL. Let me see if I can explain why I want it that way. The goal of a property based testing framework should be to create syntax which models the statement _"If the input has properties ABC, the ouput has properties XYZ"_ as accurately as possible. 

As an example, if I have a function `keep-even` that returns a list of all of the even numbers in a list, it has two pretty straightforward properties. Here is what I think the ideal syntax for describing the properties would look like:

```
(deftest keep-even

  ;; If the input is a list of integers.
  (let* ((input (generator (list (integer))))
         (output (keep-even input)))

    ;; The output should satisfy this property.
    (check-that (= (len output) (count-if #'evenp input))

    ;; And it should also satisfy this property. 
    (check-that (every #'evenp xs))))
```

Once you have a nice syntax for expressing the properties, you then need a nice way to explain what happened to the user when a property fails. The framework should be able to say, _"I have found some input with properties ABC, but the output fails to satisfy property Z"_. I am now thinking that once check-it finds some case that fails, it should shrink it with respect to the one property that initially failed, but then once it has shrunk it as much as possible, check-it should run the input against all of the other properties as well. When rerunning against all of properties, I do not see any real reason to separate the code for the properties. 

About losing the compatibility with different testing frameworks - after using check-it with clunit, I think it would be a good idea for check-it to come with its own testing framework. Most testing frameworks were not designed to handle the property based checking that check-it does and I think an entire testing framework based on the idea of properties would be a good thing. 

I'm just trying to imagine what is the best way to specify the properties, but still get good error reporting at the same time. If there is anything I forgot to mention, please let me know.

**Comment by malisper** [comment]
I managed to get a proof of concept with integrating check-it with clunit. It is extremely hacky so I may see if there is another testing framework that is easier to integrate with. Here is what using it looks like:

```
CHECK-IT> (deftest foo ()
            (with-tests
              (let* ((x (generator (integer)))
                     (y (generator (integer)))
                     (result (+ x y)))
                (check-that (= result (+ x y)))
                (check-that (= result (+ x y 1)))
                (check-that (= result (+ x y 2))))))
#<CLUNIT::CLUNIT-TEST-CASE {100933C153}>

CHECK-IT> (run-test 'foo)
Assertion (= RESULT (+ X Y 2)) failed.
Failed case: (1 -9)
Shrunken failure case: (0 0)
Testing all assertions that fail the shrunken case:
Assertion (= RESULT (+ X Y 1)) failed.
Assertion (= RESULT (+ X Y 2)) failed.

Assertion (= RESULT (+ X Y 1)) failed.
Failed case: (2 10)
Shrunken failure case: (0 0)
Testing all assertions that fail the shrunken case:
Assertion (= RESULT (+ X Y 1)) failed.
Assertion (= RESULT (+ X Y 2)) failed.


FAILURE DETAILS:
================
    FOO: Expression: (= RESULT (+ X Y 2))
         Expected: T
         Returned: NIL

    FOO: Expression: (= RESULT (+ X Y 1))
         Expected: T
         Returned: NIL


SUMMARY:
========
    Test functions:
        Executed: 1
        Skipped:  0

    Tested 3 assertions.
        Passed: 1/3 ( 33.3%)
        Failed: 2/3 ( 66.7%)
```

There are probably a couple edge cases I missed such as when an error is signaled. Right now I am going to try to improve the output and make it match the output of clunit better.

**Comment by DalekBaldwin** [comment]
I finally had a chance to come back to this, and it jumped out at me: you could handle multiple properties per test elegantly and modularly with plain old condition signaling.

No cloning and modifying blocks of code, no multiplying the number of trials by the number of `check-that` forms, no generating different sets of examples to test each property (or having to capture and reset the random state if you want to rerun the same set of examples for each property). And no need to clutter up the macroexpansions for the most common use cases just because this feature is there for power users.

Each `check-that` form can be labeled with an identifier (perhaps some gensym, behind the scenes) and it can signal a condition tagged with that identifier to indicate a specific property violation. Then if you want to minimize with respect to a single property, you run the same test body, but you bind a condition handler that ignores signals from all the other properties so they aren't treated as failures.

You could even mutate data _inside_ the individual property checks and trust that the same mutations will happen, in sequence, every time. (That would be poor style if you wrote them that way on purpose, but if you did it by accident you'd save yourself a lot of confusion in trying to interpret the test failures.)

There's only one of your pet features that this approach couldn't trivially replicate: the whole combined test would only show up as a single pass/failure from the perspective of the overlying unit test framework. But I'm convinced that's the right way to go. If you wanted to indicate that some of the other properties failed after minimizing with respect to the first property, it would be misleading to report in the official test summary that the first property failed while the other ones passed.

**Comment by DalekBaldwin** [comment]
Here's a sketch of what the overall DSL is turning into, demonstrated by combining most of the features into a single top-level form. As long as you can specify how all the low-level atomic values you need can be generated, you can specify almost any other more complex -- yet still shrinkable -- generation strategy using only mapping and chaining generators. (Some of the other built-in generators should probably be redefined in terms of a mapping generator anyway.)

``` lisp
;; sketch of lexical contextual dataflow DSL for property-based testing
(let ((x #g(integer 0 10)) ;; at local scope, define generators of integers
      (y #g(integer 4 6))) 
  (deftest blah-test () ;; define a top-level test function in some unit test
                        ;; framework
    (generating (x ;; lexically register name to same generator bound at in
                   ;; surrounding lexical scope
                 (y #g(integer 3 7)) ;; lexically register name to new generator
                                     ;; different from the one bound in
                                     ;; surrounding lexical scope
                 (z #g(integer 15 20)) ;; lexically register new generator
                 )
      (is ;; treated as single test case in whatever unit test framework you use
       (checking ;; now we actually generate individual values from
                 ;; lexically-registered generators `x`, `y`, and `z`, and
                 ;; rebind the results to the same names as the generators that
                 ;; produced them
         (that (= x y z)) ;; check/shrink this property only
         ))
      (mapping ((a (+ x y)) ;; specifies how values of `a` will be produced from
                            ;; values of `x` and `y` once we actually generate
                            ;; values within a `checking` form
                (b (+ x y z)) ;; likewise, specifies how values of `b` will be
                              ;; produced
                )
        (chaining ((w ;; chaining: generate results from surrounding generators
                      ;; in order to create parameters for this generator; then
                      ;; this generator will have its values generated through a
                      ;; separate value-generation event. (if multiple distinct
                      ;; `chaining` forms are lexically composed, we can only
                      ;; conduct a shrinkage search with respect to the nearest
                      ;; enclosing `chaining` generator and any other `mapping`
                      ;; generators that derive their values from that chained
                      ;; generator's values)
                    #g(list (integer)
                            :min-length a
                            :max-length b)))
          (is
           (checking ;; generate/rebind `x`, `y`, `z`, `a`, `b`, and `w` for
                     ;; each set of generated values...
             (that (<= a (length w) b)) ;; check this property (this one should
                                        ;; actually hold! not just a foo-bar
                                        ;; example)
             (that (= a b)) ;; check this property at the same time
             (that (something something)) ;; this one too
             )
           ;; and during shrinking, report whenever we discover a sample value
           ;; that fails a different **combination** of checked properties than
           ;; those that the last sample value failed, and issue as the final
           ;; shrunken value whatever value did not fail to satisfy any stated
           ;; properties and did not trigger any unhandled errors
           ))
        (is
         (checking ;; generate/rebind `x`, `y`, `z`, `a`, and `b` for each set
                   ;; of generated values...
           (that nil) ;; this property will always fail, obviously
           (that (= x y z a b)) ;; this will surely fail, just a BS example
           ))))))
;; the above form specifies three different test cases from the perspective of a
;; surrounding unit-test framework, with some overlapping dependencies on
;; randomized generator specifications. we take advantage of the concept of
;; lexical scope, but apply it to concepts that are not typically lexically
;; scoped, to allow users to write test specifications that are extremely
;; compact yet readable to anybody who knows how to apply compositional
;; reasoning. the specifications compose, but not in the sense of composition
;; usually applied in pure functional programming. the surrounding definitions
;; are only fully combined into a useable generating/shrinking context once you
;; enter a `checking` form.
```

Without the comments:

``` lisp
(let ((x #g(integer 0 10))
      (y #g(integer 4 6))) 
  (deftest blah-test ()
    (generating (x
                 (y #g(integer 3 7))
                 (z #g(integer 15 20)))
      (is
       (checking (that (= x y z))))
      (mapping ((a (+ x y))
                (b (+ x y z)))
        (chaining ((w #g(list (integer)
                              :min-length a
                              :max-length b)))
          (is
           (checking (that (<= a (length w) b))
                     (that (= a b))
                     (that (something something)))))
        (is
         (checking (that nil)
                   (that (= x y z a b))))))))
```

I'm not aware of any other language/framework that uses present participles to indicate a related group of linguistic constructs. (For a short while, Clojure used them for transducers (`mapping`, `filtering`, `mapcatting`, etc.), but now they've been combined with the same symbols for the corresponding higher-order functions (`map`, `filter`, `mapcat`, etc.)). I think they work pretty well here. They describe a lexical context that is continually being built up, but not actually being actively used until a specific form relies on them. I'm just not sure whether that specific form should also be a present participle (`checking`) or if it would be clearer to distinguish it some other way.
